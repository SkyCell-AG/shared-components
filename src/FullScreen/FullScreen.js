import React, {
    useRef,
    useCallback,
    useState,
    useEffect,
} from 'react'
import noop from 'lodash/noop'
import PropTypes from 'prop-types'
import clsx from 'clsx'
import FullscreenIcon from '@mui/icons-material/Fullscreen'
import FullscreenExitIcon from '@mui/icons-material/FullscreenExit'

import useStyles from './FullScreen.styles'

const propTypes = {
    className: PropTypes.string,
    children: PropTypes.element.isRequired,
    onFullScreen: PropTypes.func,
    setTemperatureChartFullscreen: PropTypes.func,
}

const defaultProps = {
    className: '',
    onFullScreen: noop,
    setTemperatureChartFullscreen: noop,
}

const FullScreen = ({
    className,
    children,
    onFullScreen,
    setTemperatureChartFullscreen,
}) => {
    const classes = useStyles()
    const wrapper = useRef(null)
    const [
        fullScreen,
        setFullScreen,
    ] = useState(false)

    const requestFullScreen = useCallback(() => {
        let prevented = false
        const event = {
            preventDefault: () => {
                prevented = true
            },
        }

        onFullScreen(event)
        if (prevented) {
            return
        }

        if (fullScreen) {
            document.exitFullscreen()
        } else {
            wrapper.current.requestFullscreen({
                navigationUI: 'show',
            }).then({})
        }
    }, [
        fullScreen,
        onFullScreen,
    ])

    const onFullScreenChange = useCallback(() => {
        setFullScreen(document.fullscreenElement)
        setTemperatureChartFullscreen(document.fullscreenElement)
        if (document.fullscreenElement) {
            document.body.classList.add(classes.fullscreen)
            const chartContainer = document.querySelector(`.${classes.children}`)

            chartContainer.style.display = 'flex'
            chartContainer.style.alignItems = 'center'
            chartContainer.style.height = '100%'
        } else {
            document.body.classList.remove(classes.fullscreen)

            const chartContainer = document.querySelector(`.${classes.children}`)

            chartContainer.style.display = ''
            chartContainer.style.alignItems = ''
            chartContainer.style.height = ''
        }
    }, [
        classes.children,
        classes.fullscreen,
        setTemperatureChartFullscreen,
    ])

    useEffect(() => {
        const currentWrapper = wrapper.current

        currentWrapper.addEventListener('fullscreenchange', onFullScreenChange)
        return () => {
            currentWrapper.removeEventListener('fullscreenchange', onFullScreenChange)
        }
    }, [onFullScreenChange])

    return (
        <div
            className={clsx(
                className,
                classes.wrapper,
            )}
            ref={wrapper}
        >
            <div
                className={fullScreen
                    ? classes.fullScreenButtonClose
                    : classes.fullScreenButtonOpen}
                onMouseDown={requestFullScreen}
            >
                {fullScreen ? (
                    <FullscreenExitIcon />
                ) : (
                    <FullscreenIcon />
                )}
            </div>
            <div
                className={classes.children}
            >
                {children}
            </div>
        </div>
    )
}

FullScreen.propTypes = propTypes
FullScreen.defaultProps = defaultProps

export default FullScreen
