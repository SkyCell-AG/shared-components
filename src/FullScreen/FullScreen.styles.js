import {
    makeStyles,
} from '@material-ui/core'

const useStyles = makeStyles((theme) => {
    return {
        wrapper: {
            position: 'relative',
            height: '100%',
            width: '100%',
            paddingTop: '2rem',
            alignItems: 'center',
            background: theme.palette.common.white,
        },
        fullScreenButtonClose: {
            top: 20,
            right: 20,
            zIndex: 1,
            position: 'absolute',
            opacity: 0.2,
            cursor: 'pointer',
        },
        fullScreenButtonOpen: {
            zIndex: 1,
            position: 'absolute',
            right: 0,
            top: '-2rem',
            opacity: 0.2,
            cursor: 'pointer',
        },
        children: {
            height: '100%',
            width: '98%',
        },

    }
})

export default useStyles
