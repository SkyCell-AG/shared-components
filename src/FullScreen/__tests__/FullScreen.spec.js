import React from 'react'
import {
    shallow,
} from 'enzyme'

import FullScreen from '../FullScreen'

describe('Apps/shared-components/FullScreen', () => {
    it('FullScreen match snapshot', () => {
        const Children = () => {
            return (<div>children</div>)
        }

        const wrapper = shallow(
            <FullScreen
                className="makeStyles-wrapper-244"
                onFullScreen={jest.fn()}
                setTemperatureChartFullscreen={jest.fn()}
            >
                <Children />
            </FullScreen>,
        )

        expect(wrapper).toMatchSnapshot()
    })
})
