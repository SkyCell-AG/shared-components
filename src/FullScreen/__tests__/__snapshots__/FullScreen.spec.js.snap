// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Apps/shared-components/FullScreen FullScreen match snapshot 1`] = `
<div
  className="makeStyles-wrapper-244 makeStyles-wrapper-1"
>
  <div
    className="makeStyles-fullScreenButtonOpen-3"
    onMouseDown={[Function]}
  >
    <Memo(ForwardRef(FullscreenIcon)) />
  </div>
  <div
    className="makeStyles-children-4"
  >
    <Children />
  </div>
</div>
`;
